### PR TITLE
[SPARK-5115] specify hadoop version for yarn module deps

### PR DESCRIPTION
**Pull request by ryan-williams**
fixes IntelliJ's dependency-resolution in the yarn module

NOTE: #3917 is my preferred fix for this issue, but I've submitted this one for comparison

### CODE REVIEW

File: yarn/pom.xml [hunk 1]
```diff
@@ -41,28 +42,34 @@
     <dependency>
       <groupId>org.apache.hadoop</groupId>
       <artifactId>hadoop-yarn-api</artifactId>
+      <version>${hadoop.version}</version>
```
Comment (srowen) [inline yarn/pom.xml]: This should be `yarn.version` if anything, but should not be necessary. The version is managed in the parent POM. This is not the right way to do this.

Comment (ryan-williams) [inline yarn/pom.xml]: right, `yarn.version`.

I agree that this feels like a worse route than #3917, as I've mentioned in a few places.

OOC, would the `yarn.version` (or `hadoop.version` -> `yarn.version`) that a given profile chooses override a `yarn.version` that was set here? I am guessing so, but am not sure.

**Comment by srowen** [comment]
A `<version>` here would override what's in the `spark-parent` POM. So this would force it to `hadoop.version`. `yarn.version` itself defaults to take the value of `hadoop.version`, but is separate so that they could be set separately back in the days when the versions didn't track each other. A system property would override values set in the POM's profiles.

**Comment by ryan-williams** [comment]
Interesting. So the profiles in the parent POM will override the default `hadoop.version` there (and set `yarn.version` to the `hadoop.version` only _after_ the profiles have had a chance to override `hadoop.version`), but they will not override defaults that are set in child POMs? So the order is roughly:

1. set `hadoop.version` to `1.0.4` by default (parent POM)
1. let a specific profile override this default (parent POM)
1. set `yarn.version <- hadoop.version`
1. let child POMs further override `{hadoop,yarn}.version`
1. go fetch appropriate dependencies based on last values of `{hadoop,yarn}.version`

?

**Comment by srowen** [comment]
Well right now versions are managed in the parent POM, on purpose. A declaration in the child would override a declaration in the parent, if it existed. Children don't behave differently with respect to precedence of system properties set on the command line, and profiles. Child POMs do not (in general) manage versions like this, and shouldn't. But yes 1-3 looks right. Step 2.5 I believe would be to let `-Dhadoop.version=...` override these values too.

**Comment by ryan-williams** [comment]
makes sense. Looking again at the POMs, it seems like I need to revise my mental model a little further.

Step 3 above (`yarn.version` assigned to value of `hadoop.version`) happens [right alongside](https://github.com/apache/spark/blob/bb38ebb1abd26b57525d7d29703fd449e40cd6de/pom.xml#L122-L124) Step 1 (`hadoop.version` given a default of `1.0.4`).

Based on that, it seems like the process must be more like:

1. set properties based on `-Dhadoop.version=...`-style system args.
1. set _any remaining unset properties_ based on `<profile>` declarations.
1. set _further remaining unset properties_ based on the defaults in `<properties>`.

Then I can see a few ways in which it might handle parsing child POMs and incorporating values found there at different steps that would result in the wrong behavior happening here.

In any case, I'll close this out and get back to you on the JIRA, unless you have any other ruminations to offer about Maven-prop-value-resolution arcana. Thanks!

**Pull request closed without merging**